### PR TITLE
[CI] Build PR images on GHCR

### PR DESCRIPTION
**Pull request by jlestel**
I modified the CI so that the PR images are build then pushed into the GitHub packages (GHCR).

This does not interfere with how current CI works with Docker Hub.

Functioning :
- When a PR is created or modified: the image is built on the GHCR from fork repo.
- When a branch is pushed: the image is built on the GHCR from main repo.
- When a PR is completed: the image is deleted in the fork repo. Additionally, untagged images are also deleted on this fork repo.

Once this PR is merged, each new / updated PR will trigger CI on the fork repo.
- Users will then see all these images on their fork. Eg: [https://github.com/jlestel/teslamate/pkgs/container/teslamate](https://github.com/jlestel/teslamate/pkgs/container/teslamate)

If necessary, anyone can then modify their docker-compose to use a PR image (eg: `ghcr.io/jlestel/teslamate:master`) for testing this PR instead of `teslamate/teslamate:latest`.

### CODE REVIEW

**Comment by jlestel** [comment]
> Thank you for this PR. Can you elaborate on why you chose to modify the Elixir CI rather than create your own workflow?

No reason, I just separated them.

Now I trigger the GHCR build on a pull_request or push master in a separate workflow.

File: .github/workflows/ghcr_build.yml
Comment (JakobLichterfeld) [inline .github/workflows/ghcr_build.yml:9]: I would add a `workflow_dispatch:`as well

File: .github/workflows/ghcr_build.yml
Comment (JakobLichterfeld) [inline .github/workflows/ghcr_build.yml:19]: For maintenance reasons, I would extract the buildx to a separate action since we now use it in different places.

File: .github/workflows/ghcr_build.yml
Comment (JakobLichterfeld) [inline .github/workflows/ghcr_build.yml:35]: I would add a timeout to prevent stuck jobs

**Review by JakobLichterfeld — CHANGES_REQUESTED** [review]
lgtm, my findings see inline comments

File: .github/workflows/ghcr_build.yml
Comment (jlestel) [reply .github/workflows/ghcr_build.yml:10]: Added

File: .github/workflows/ghcr_build.yml
Comment (jlestel) [reply .github/workflows/ghcr_build.yml:37]: Done

**Comment by DrMichael** [comment]
> If necessary, anyone can then modify their docker-compose to use a PR image (eg: `ghcr.io/teslamate-org/teslamate:pr-123`) for testing instead of `teslamate/ teslamate:latest`

Would that also be the case for grafana?

File: .github/workflows/ghcr_build.yml
Comment (jlestel) [reply .github/workflows/ghcr_build.yml]: I tried to refactor things. Let me know what you think about.

And I believe the Docker Hub user has changed since my initial commit.

**Comment by jlestel** [comment]
> > If necessary, anyone can then modify their docker-compose to use a PR image (eg: `ghcr.io/teslamate-org/teslamate:pr-123`) for testing instead of `teslamate/ teslamate:latest`
> 
> Would that also be the case for grafana?

yes I added grafana also

**Comment by adriankumpf** [comment]
Could this be used by a malicious PR to leak build secrets? Or does building the images require maintainer approval?

**Comment by brianmay** [comment]
I don't like to be the pessimist, but wondering if this will work. "With the exception of GITHUB_TOKEN, secrets are not passed to the runner when a workflow is triggered from a forked repository." https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions#using-secrets-in-a-workflow

My experience, not even jobs triggered by dependabot created pull requests have access to secrets. I had to update all of my github workflows as a result.

But nothing wrong with trying it. :-)

Also see: https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions

**Comment by JakobLichterfeld** [comment]
> Could this be used by a malicious PR to leak build secrets? Or does building the images require maintainer approval?

Good point you raise. As far as my brief research shows, you can extract secrets using malicious github workflows and actions. Building a docker image with a verified pipeline should not be an issue, but PRs related to .github should be excluded from auto-run. But this is nothing introduced by this PR, but more of a general issue in my opinion.

File: .github/actions/build/action.yaml
Comment (JakobLichterfeld) [inline .github/actions/build/action.yaml:62]: It should be /tmp/teslamate/digests/*

File: .github/actions/grafana/action.yml
Comment (JakobLichterfeld) [inline .github/actions/grafana/action.yml:14]: I would remove the myteslamate :-)

File: .github/actions/grafana/action.yml
Comment (JakobLichterfeld) [inline .github/actions/grafana/action.yml:29]: Can we really do without it?

```
 - name: Login to DockerHub
        uses: docker/login-action@v3.0.0
        with:
          username: teslamate
          password: ${{ secrets.DOCKER_PASSWORD }}
```

File: .github/actions/merge/action.yml
Comment (JakobLichterfeld) [inline .github/actions/merge/action.yml:10]: Shouldn't it be /tmp/teslamate/digests?

File: .github/actions/merge/action.yml
Comment (JakobLichterfeld) [inline .github/actions/merge/action.yml:11]: Separation of concerns: Is setup buildx a merge action?

**Review by JakobLichterfeld — CHANGES_REQUESTED** [review]
hank you for your dedicated work! I really like the process of refactoring into separate actions. I added my findings as inline comments.

**Review by JakobLichterfeld — COMMENTED** [review]
lgtm

File: .github/actions/build/action.yaml
Comment (jlestel) [reply .github/actions/build/action.yaml]: I finished to refactoring. We finally stay on /tmp/digest.

File: .github/actions/grafana/action.yml
Comment (jlestel) [reply .github/actions/grafana/action.yml:29]: It seems to work (I do it in the workflow before this action)

File: .github/actions/merge/action.yml
Comment (jlestel) [reply .github/actions/merge/action.yml]: Reverted

**Comment by jlestel** [comment]
> Could this be used by a malicious PR to leak build secrets? Or does building the images require maintainer approval?

Yes there was a flaw there.

Now, I prevent workflows from being launched when there are changes to the .github directory: only a maintainer can run it manually.

**Comment by jlestel** [comment]
> I don't like to be the pessimist, but wondering if this will work. "With the exception of GITHUB_TOKEN, secrets are not passed to the runner when a workflow is triggered from a forked repository." https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions#using-secrets-in-a-workflow
> 
> My experience, not even jobs triggered by dependabot created pull requests have access to secrets. I had to update all of my github workflows as a result.
> 
> But nothing wrong with trying it. :-)
> 
> Also see: https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions

You're right. In any case, secrets.GITHUB_TOKEN and secrets.DOCKER_PASSWORD will not be accessible to PRs from external repos... That seems normal to me too: it is rather work branches or release candidates whose content has been qualified by maintainers that could be proposed.

This involves modifying the gitflow to no longer use only a master branch and be able to create GH images on these new branches.

Maybe you already have ideas on a gitflow that could make sense?

**Comment by jlestel** [comment]
> The good news: it should work with secrets.GITHUB_TOKEN and this is the only one we should use.

As you can see [in this action](https://github.com/teslamate-org/teslamate/actions/runs/6891889827/job/18747960155), the GITHUB_TOKEN is in read only.

This seems linked to the [Maximum access for pull requests from public forked repositories](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token) which limit write access even if permissive mode is on.

So, i'll try to replace pull_request by pull_request_target to only run an action in the repo of the requester of a PR.

**Comment by jlestel** [comment]
> So, i'll try to replace pull_request by pull_request_target to only run an action in the repo of the requester of a PR.

Voilà @adriankumpf @JakobLichterfeld : the GHCR build workflow is now based on the PR source user. The CI is played on the fork repo  instead of the teslamate-org repo (and the generated images too).

Eg for this PR, you'll find image(s) on my fork:
- TeslaMate image: https://github.com/jlestel/teslamate/pkgs/container/teslamate
- Grafana image: https://github.com/jlestel/teslamate/pkgs/container/teslamate

It build only amd64 as others need a thrid party not everyone has install.

Let me know what you think about that.

File: .github/workflows/ghcr_build.yml
Comment (JakobLichterfeld) [inline .github/workflows/ghcr_build.yml]: Please remove your debug code. It may look suspicious to a casual observer.

File: .github/actions/merge/action.yml
Comment (JakobLichterfeld) [inline .github/actions/merge/action.yml:30]: What about a Login to the docker hub?
```      - name: Login to Docker Hub
        uses: docker/login-action@v3.0.0
```

File: .github/workflows/buildx.yml
Comment (JakobLichterfeld) [inline .github/workflows/buildx.yml:55]: I see this as a requirement inside the merge action

File: .github/workflows/ghcr_build.yml
Comment (JakobLichterfeld) [inline .github/workflows/ghcr_build.yml:26]: please enable all platforms

File: .github/workflows/ghcr_build.yml
Comment (JakobLichterfeld) [inline .github/workflows/ghcr_build.yml:29]: please enable all platforms

File: .github/workflows/buildx.yml
Comment (JakobLichterfeld) [reply .github/workflows/buildx.yml:55]: Forget it, I've got it now. Is outsourced to separate ghcr and docker hub push.

File: .github/actions/merge/action.yml
Comment (JakobLichterfeld) [reply .github/actions/merge/action.yml:30]: Forget it, I've got it now. Is outsourced to separate ghcr and docker hub

**Review by JakobLichterfeld — CHANGES_REQUESTED** [review]
Thanks again for refactoring and your dedicated work. 

I added once again my small findings.

Additionally, two general points still prevent me from merging:
- I would vote to include ```- '!.github/**'  # Important: Exclude PRs related to .github from auto-run``` in all workflows
- Please lint the files, as there are some trailing whitespaces for example

File: .github/workflows/ghcr_build.yml
Comment (jlestel) [reply .github/workflows/ghcr_build.yml]: Outdated: was in WIP at this moment

File: .github/workflows/ghcr_build.yml
Comment (jlestel) [reply .github/workflows/ghcr_build.yml]: When I was testing. I re-added it.

File: .github/workflows/ghcr_build.yml
Comment (jlestel) [reply .github/workflows/ghcr_build.yml]: done. forkers who want to build will have to use the buildjet service

**Review by JakobLichterfeld — APPROVED** [review]
lgtm

**Comment by JakobLichterfeld** [comment]
Merging on mobile in the middle of the night without testing in our repo was not the best idea. I suggest opening a CI branch, including the branch in the workflow triggers, and testing it in our repo, as local testing of github workflows is limited in terms of secrets and github usernames.

**Comment by jlestel** [comment]
> @jlestel Sorry, merging on mobile in the middle of the night without testing in our repo was not the best idea. I suggest opening a CI branch, including the branch in the workflow triggers, and testing it in our repo, as local testing of GitHub workflows is limited in terms of secrets and GitHub usernames. The issue lies within the login to pull buildx.

No problem. Yes it will be better to test it with a dedicated branch as well as on our repositories. Keep me informed.

**Pull request merged**
